fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/BoundingBoxOcclusionModifier.java

```diff
@@ -0,0 +1,172 @@
+    /**
+     * This defines how many equidistant points shall be evaluated per edge of a vehicle.
+     * Note generally for the front and rear edge this will result in a higher resolution compared to the sides of the vehicle.
+     * Default: 2
+     */
+    private final int pointsPerSide;
+
+    /**
+     * Threshold that defines how many of the points defined through {@link #pointsPerSide} need to be visible in order for a
+     * object to be treated as detected.
+     * Default: 2
+     */
+    private final int detectionThreshold;
+
+    /**
+     * Default constructor for the {@link BoundingBoxOcclusionModifier}.
+     */
+    public BoundingBoxOcclusionModifier() {
+        this.pointsPerSide = 2;
+        this.detectionThreshold = 2;
+    }
+
+    /**
+     * Constructor for {@link BoundingBoxOcclusionModifier}, validates and sets
+     * the parameters {@link #pointsPerSide} and {@link #detectionThreshold}.
+     *
+     * @param pointsPerSide      the amount of points that will be evaluated per object (corners count towards 2 edges)
+     * @param detectionThreshold how many points have to be visible in order for an object to be treated as detected
+     */
+    public BoundingBoxOcclusionModifier(int pointsPerSide, int detectionThreshold) {
+        if (pointsPerSide < 2) {
+            throw new RuntimeException("Need at least 2 points per edge, meaning every corner will be checked for occlusion.");
+        }
+        if (detectionThreshold < 1) {
+            throw new RuntimeException("At least one point has to be checked for occlusion, else no objects will be occluded");
+        }
+        if (detectionThreshold > pointsPerSide * 4 - 4) {
+            throw new RuntimeException("The detection threshold exceeds the number of points evaluated per object");
+        }
+        this.pointsPerSide = pointsPerSide;
+        this.detectionThreshold = detectionThreshold;
+    }
+
+    @Override
+    public <T extends SpatialObject> List<T> apply(PerceptionModuleOwner owner, List<T> spatialObjects) {
+        List<T> newObjects = new ArrayList<>();
+        // the ego object cannot occlude vision
+        List<T> occludingObjects = spatialObjects.stream()
+                .filter(object -> !object.getId().equals(owner.getId()))
+                .collect(Collectors.toList());
+        for (T spatialObject : spatialObjects) {
+            if (spatialObject instanceof TrafficLightObject) { // Traffic Lights are treated to not be occluded
+                newObjects.add(spatialObject);
+                continue;
+            }
+            List<Vector3d> pointsToEvaluate = createPointsToEvaluate(spatialObject);
+            final int requiredVisiblePoints = pointsToEvaluate.size() == 1 ? 1 : detectionThreshold;
+            int numberOfPointsVisible = 0;
+            for (Vector3d point : pointsToEvaluate) {
+                boolean pointOccluded = false;
+                boundingBoxLoop:
```

Comment:
instead of using labels, try to extract the loop into a new method, e.g. `boolean pointOccluded = isPointOccluded(...) `

Reply:
I extracted the method and also inverted the boolean logic to make the code more readable